### ISSUE TITLE
Suggestion: rush() — eagerly takes n first elements of an iterable an leaves the rest

### ISSUE THREAD

**Issue by KSXGitHub** [sithmel/iter-tools#136]
## Example

```javascript
const [rushed, rest] = rush(4, range())
```

**Expect:**
  * `rushed` should be an array of 4 numbers: `[0, 1, 2, 3]`
  * `rest` should be an iterator starting from 4

**Comment by conartist6** [comment]
Could you provide a potential use case?

**Comment by KSXGitHub** [comment]
> Could you provide a potential use case?

The example is a potential use case.

User may want some elements early on (`rust`) and pass the rest (`rest`) to some other functions.

**Comment by KSXGitHub** [comment]
> What you've given isn't a use case.

I honestly don't know what you demand of me. I have described the only use case I had in mind.



> Why would I write that? What would I be trying to get done? Why would I not just write:
> 
> ```js
> const firstNumbers = Array.from(range(1, x));
> const restNumbers = range(x);
> ```

`range()` in my example is just an example iterable, I could've just replace it with the word `iterable`.

And also, your suggested alternative is incorrect, `range(x)` does not go from x to ∞ but goes from 0 to x, so it does not look like `rest` in my example at all.



> Look at it this way: whenever we make a new name for something, it has to be worth it for most people to learn the name instead of just using (more or less) the implementation inline. To meet that criteria one of the following must be the case:
> 
> * The need for the implementation is very common.
> * The implementation is complicated and/or bug-prone.
> * The name is so good that people can understand the implementation without additional help.

I'm bad at naming things, especially in abstract situation like this 🙂. If anyone knows a better name, I'm all ears.

**Comment by conartist6** [comment]
I have amended my example to use `range({start: 3})`.

In general my thinking is that in most situations the need to do this would either indicate bad data structuring.

For example I would not want to see:
```js
const bigCats = ["Cecil", "Aslan", "Hobbes", "Richard Parker"];
const [lions, tigers] = rush(2, bigCats);
```

I'd rather see:
```js
const bigCats = {
   lions: ["Cecil", "Aslan"],
   tigers: ["Hobbes", "Richard Parker"],
}
printCatNames(bigCats.lions); // lions only
printCatNames(bigCats.tigers); // tigers only 
printCatNames(flat(values(bigCats)); // all cats
```

**Comment by KSXGitHub** [comment]
@conartist6 What's about recursive functions?

```javascript
function doSomething (iterable) {
  const [[a, b], rest] = rush(2, iterable)
  ... // do something with a, b, and rest
  return doSomething(chain(b, rest))
}
```

**Comment by sithmel** [comment]
@conartist6 I will adopt your bullet points as guideline https://github.com/sithmel/iter-tools/wiki/Function-guidelines

@KSXGitHub I can see what you want to get here. I am not sure it is worth adding it to iter-tools. Let me think about it

**Comment by KSXGitHub** [comment]
@sithmel BTW, there's a few `partition` related pull requests that have yet to be settled, I'm not confident enough to start working on async versions right now until you make your decisions.

**Comment by conartist6** [comment]
@KSXGitHub I like that example better, but it what you've written is pretty much just the array destructuring operator, which already exists and works on iterables. It does turn rest into an array though.

**Comment by conartist6** [comment]
```js
const [a, b, ...rest] = (function *(){yield 1; yield 2; yield 3; yield 4;})();
// a: 1;
// b: 2;
// rest: [3, 4];
```

**Comment by KSXGitHub** [comment]
> what you've written is pretty much just the array destructuring operator, which already exists and works on iterables

Yes, but `rush` does not iterate over `rest`, and that's the point.

**Comment by conartist6** [comment]
It could just be called "destructure". That at least would satisfy the critera that it be a name from which people can (relatively) easily, infer the implementation of the function.

I guess I'm still looking for a use case. Most of the time I would not want an array in which the first few items were any different than all the rest, as in my big cats example. The one use case I can think of is varArgs functions, which I imagine accounts for the majority of usages of array (ok, argument really) rest spreads in javascript. The thing is, varArgs functions, while they may have a leading parameter or two (as do many in iter-tools) are already handled by the existing syntaxes.

**Comment by KSXGitHub** [comment]
> It could just be called `destructure`.

It doesn't really describe the behavior to me, and it vaguely describes one of many possible use-cases.

You can use more than one words, such as `rushAndWait` (I'm bad at naming).

**Comment by sithmel** [comment]
Is this equivalent to rush ?
```js
const [iter1, iter2] = tee(iterable)
const a = slice(4, iter1)
const b = slice({ start: 4 }, iter2)
```

**Comment by KSXGitHub** [comment]
> Is this equivalent to rush ?
> 
> ```js
> const [iter1, iter2] = tee(iterable)
> const a = slice(4, iter1)
> const b = slice({ start: 4 }, iter2)
> ```

Equivalent on the surface, not so much under the hook as both `a` and `b` are lazy in this version.

Beside, you wrote more code (3 lines as opposed to just 1), and you used more variables (iter1, iter2).

**Comment by sithmel** [comment]
> Beside, you wrote more code, you used more variables, and it looks more cluttered.

**tee** is designed to cover most of cases where you need to return multiple iterable . I wonder if instead of trying to solve this specific case, I should try figuring out a way to get a **tee** that is less cumbersome.
```js
const a = toArray(slice(4, fork(iterable)))
const b = slice({ start: 4 }, fork(iterable))
```

**Comment by conartist6** [comment]
> It doesn't really describe the behavior to me, and it vaguely describes one of many possible use-cases.

I still haven't heard a single real use case. Argument destructuring would be one if it wasn't already handled by spread, and in general that pattern is not one that is replicated elsewhere. I want to be clear that I'm not saying that we should never add such a method. I'm just saying that we shouldn't add it before we even know what people are going to be trying to accomplish with it. We should add methods in response to real needs arising in real code when we cannot recommend any better solution.

As to creating some more powerful syntax for tee, it's an interesting thought.

**Comment by KSXGitHub** [comment]
> No idea on how to implement "proxy" though

If I was to guess, this "proxy" creates an iterator that yields values from given iterable without consuming said iterable?

**Comment by conartist6** [comment]
I feel like `tee` is still the right function to use. More importantly, `tee` is properly functional, while an implementation of `proxy` clearly is not, given the way that it is used.

The code for the proposed function rush would currently be written:
```js
const [iter1, iter2] = tee(iterable)
const a = Array.from(slice(4, iter1))
const b = slice({ start: 4 }, iter2)
```
Actually now that I think about it I believe I know how to do this cleanly without a rush method.
How about the following?
```js
const [[...a], b] = splitAt(idx, iterable);
```

**Comment by KSXGitHub** [comment]
> How about the following?
> 
> ```js
> const [[...a], b] = splitAt(idx, iterable);
> ```

While I do like the elegance of `splitAt`, it would require a cache (or `tee()`) to support iterators.

**Comment by conartist6** [comment]
Yeah, indeed it would. It could use a simple array cache though since it would know right up front how many elements it was going to need to cache, and it could avoid making the array cache at all if no items were consumed from the latter half of the split before the iterable for the first half of the split was exhausted, which would be the case in the sample code above.

**Comment by KSXGitHub** [comment]
So I guess I should reopen https://github.com/sithmel/iter-tools/issues/134 then.

**Comment by sithmel** [comment]
closed in favour of #134
